conbench/entities/history.py

```diff
@@ -85,31 +89,170 @@ def get_history(case_id, context_id, hardware_hash, repo) -> List[tuple]:
             Hardware.hash == hardware_hash,
-            Distribution.case_id == case_id,
-            Distribution.context_id == context_id,
-            Distribution.hardware_hash == hardware_hash,
         )
-        .order_by(Commit.timestamp.asc())
-        .all()
+        .subquery()
     )
+
+    # Add column with cumulative sum of distribution changes
+    history = Session.query(
+        history,
+        func.sum(history.c.begins_distribution_change.cast(s.Integer))
+        .over(order_by=history.c.timestamp)
+        .label("segment_id"),
+    ).subquery()
+
+    # Add column with rolling mean of the means (only inside of the segment)
+    history = Session.query(
+        history,
+        func.avg(history.c.mean)
+        .over(
+            partition_by=history.c.segment_id,
+            order_by=history.c.timestamp,
+        )
+        .label("rolling_mean_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.coalesce(
+            func.avg(history.c.mean).over(
+                partition_by=history.c.segment_id,
+                order_by=s.extract("epoch", history.c.timestamp),
+                range_=(None, -1),
+            ),
+            history.c.mean,
+        ).label("rolling_mean_excluding_this_commit"),
+    ).subquery()
+
+    # Add column with the residuals from the exclusive rolling mean
+    history = Session.query(
+        history,
+        (history.c.mean - history.c.rolling_mean_excluding_this_commit).label(
+            "residual"
+        ),
+    ).subquery()
+
+    # Add column with the rolling standard deviation of the residuals
+    # (these can go outside the segment since we assume they don't change much)
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(order_by=history.c.timestamp)
+        .label("rolling_sd_including_this_commit"),
+    ).subquery()
+
+    # Add same column, excluding the current commit
+    history = Session.query(
+        history,
+        func.stddev(history.c.residual)
+        .over(
+            order_by=s.extract("epoch", history.c.timestamp),
+            range_=(None, -1),
+        )
+        .label("rolling_sd_excluding_this_commit"),
+    )
+
+    return history.all()
+
+
+def _calc_z_score(
+    data_point: float, unit: str, dist_mean: Optional[float], dist_sd: Optional[float]
+) -> Optional[float]:
+    """Calculate the z-score of a data point compared to a distribution."""
+    if dist_mean is not None and dist_sd is not None and dist_sd != 0:
+        z_score = (data_point - dist_mean) / dist_sd
+    else:
+        z_score = None
+
+    if z_score and _less_is_better(unit):
+        z_score = z_score * -1
+
+    return z_score
+
+
+def append_z_scores(benchmark_results: List[BenchmarkResult]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
```

Comment:
"append" in the function name vs "populate": can we use the same verb, whichever is more appropriate?

Could we add to this docstring

- explicitly that this tries to write to the database? 
- a recommendation for when this should be called / how often, and/or the disadvantages of not calling this?

Reply:
>append vs populate vs set

I can't find the GitHub PR comment, but I remember there was some confusion in the past when it was called `set_z_scores()`, because that name sounded like z-scores were being permanently stored in the database. That's not the case in this function; the attribute is just being set on the Python object, but it's not being saved into the DB. I'll definitely clarify that in the docstring. I think the verb `append` might be a little clearer in that regard than `set`.

Reply:
> explicitly that this tries not to write to the database? via sqlalchemy magic often times we have implicit write-to-db in the code base (where it isn't maybe immediately obvious), so if this should deliberately not write to the DB, it might be worth pointing that out.

💯 on this point. sqlalchemy is fantastic in the way it just takes care of many of these things, but coming back and trying to maintain something + remember if it's using sqlalchemy's magic (I guess that would be just alchemy!) or not can be really hard | frustrating

Reply:
> I think the verb append might be a little clearer in that regard than set.

"Append" is a rather specific verb in programming which implies that a value is appended to some kind of ordered container object. That is, doing "append" multiple times usually adds elements, retaining the previous ones. When I saw "append" it told me that when I call `append_z_scores(input)` twice on the same `input` that the result is different the second time than it was the first time.

IMO when one says "set the `z_score` attribute on the Python object" then that's a precise, explicit description of the behavior using established terminology. But even then it makes sense to add a statement about the intended behavior when calling this more than once for the same input. That is why I asked about overwrite behavior, and how you would specify it. Should this skip (to save time) when it sees an object that has this attribute already set? Should it error out to help reveal programming errors? Should it  re-calculate and replace the previous value?

It's not too important to further change this. 

I just wanted to clarify where I am coming from when I _read_ code and try to understand the interface of this function, the intended behavior.